FIAT/hdiv_trace.py

```diff
@@ -49,47 +53,97 @@ class HDivTrace(FiniteElement):
 
-        # Construct entity ids (assigning top. dim. and initializing as empty)
+            else:
+                # If one degree is provided, we splat the degree
+                # over all component cells
+                spanning_degrees = (degree,) * len(ref_el.cells)
+        else:
+            assert isinstance(ref_el, (Simplex, FiredrakeQuadrilateral))
+
+            # Cannot have varying degrees for these reference cells
+            assert not isinstance(degree, tuple), (
+                "Must have a tensor product cell if providing multiple degrees"
+            )
+            spanning_degrees = (degree,)
+
+        # Initialize entity dofs and construct the DG elements
+        # for the facets
+        facet_sd = sd - 1
+        dg_elements = {}
         entity_dofs = {}
-
-        # Looping over dictionary of cell topology to construct the empty
-        # dictionary for entity ids of the trace element
         topology = ref_el.get_topology()
-        for top_dim, entities in topology.items():
+        for top_dim, entities in iteritems(topology):
+            cell = ref_el.construct_subelement(top_dim)
             entity_dofs[top_dim] = {}
+
+            # We have a facet entity!
+            if cell.get_spatial_dimension() == facet_sd:
+                dg_elements[top_dim] = construct_dg_element(cell,
+                                                            spanning_degrees)
+            # Initialize
             for entity in entities:
                 entity_dofs[top_dim][entity] = []
 
-        # Filling in entity ids and generating points for dual basis
-        nf = dglagrange.space_dimension()
-        points = []
-        num_facets = sd + 1
-        for f in range(num_facets):
-            entity_dofs[sd - 1][f] = range(f * nf, (f + 1) * nf)
+        # Compute the dof numbering for all facet entities
+        # and extract nodes
+        offset = 0
+        pts = []
+        for facet_dim in sorted(dg_elements):
+            element = dg_elements[facet_dim]
+            nf = element.space_dimension()
+            num_facets = len(topology[facet_dim])
 
-            for dof in dglagrange.dual_basis():
-                facet_point = list(dof.get_point_dict().keys())[0]
-                transform = ref_el.get_entity_transform(sd - 1, f)
-                points.append(tuple(transform(facet_point)))
+            for i in range(num_facets):
+                entity_dofs[facet_dim][i] = range(offset + i * nf,
+                                                  offset + (i + 1) * nf)
+
+                # Run over nodes and collect the points for point evaluations
+                for dof in element.dual_basis():
+                    facet_pt = list(dof.get_point_dict().keys())[0]
+                    transform = ref_el.get_entity_transform(facet_dim, i)
+                    pts.append(tuple(transform(facet_pt)))
+
+            offset += entity_dofs[facet_dim][num_facets - 1][-1] + 1
```

Comment:
The first entry matches in the two series, and you do not actually use the last update, so it accidentally still works when `len(dg_elements) <= 2`.